### ISSUE TITLE
Partial type-in-type 

### ISSUE THREAD

**Issue by identicalsnowflake** [agda/agda#2295]
`Play/TINT.agda`:
```agda
{-# OPTIONS --type-in-type #-}

module Play.TINT where

open import Data.Container

module MyModule {ℓ} (c : Container ℓ) where
```

`Play/Boom.agda`:
```agda
module Play.Boom where

open import Play.TINT
open import Data.Container

data ⊤ : Set where
  tt : ⊤

c : Container _
c = ⊤ ▷ λ _ → ⊤

open MyModule c

```

Trying to compile `Boom` yields the error `Set₁ != Set when checking that the expression c has type Container .Agda.Primitive.lzero`, but if you C-c C-d, you can clearly see that the expression `c` *does* have type `Container .Agda.Primitive.lzero`.

**Comment by andreasabel** [comment]
With `--type-in-type`, `Set` has sort `Set`, but without it, `Set` has sort `Set1`.  I suspect this is the reason for the error.

What do you expect if you mix `--type-in-type` modules with predicative ones?  What job should `--type-in-type` do on sort annotations to allow uses without `--type-in-type`?  Note that with `--type-in-type`, all sorts are equal, so you can (by specification) only expect random sorts in the exported code.

Would you assume the following to work?

``` agda
{-# OPTIONS --type-in-type #-}

module TypeInType

A : Set
A = Set₁
```

and

``` agda
module ImpTypeInType where

open import TypeInType

B : A
B = Set
```

?
`A = Set1` so `Set : A` should work?!
Agda complains

```
Set₂ != Set
when checking that the expression Set has type A
```

Is Agda justified to complain here?
Well `B` has type `Set1` of sort `Set2`, while `A` is `Set1` of sort `Set` thus, the error.

If you turn on `--type-in-type`, then any type has sort `Set`, including `Set` itself..

If you have a complete story of how mixing of `--type-in-type` with ordinary modules should work, I would be eager to learn...

**Comment by identicalsnowflake** [comment]
I'm only just now experimenting with type-in-type (because of this interpreter thing), so it's possible I'm simply not understanding it right and it's not what I need. I was hoping rather than actually have truly 1 inconsistent universe, that it would still have all the universes and just let me put stuff in them as I please: in this case, putting the interpreter in the same universe as the `t` parameter.

Basically `primTrustMe` for universes.

**Comment by jespercockx** [comment]
You may find this interesting: http://www.cs.bham.ac.uk/~mhe/impredicativity-via-rewriting/resize.html#1

**Comment by identicalsnowflake** [comment]
@jespercockx yes, that looks like it might be what I need! I've never used `REWRITE` before. That `resize` should just change the type, right? And not affect the reduction behavior of the actual term?

**Comment by jespercockx** [comment]
Yes, the resize-id rewrite rule says exactly that `resize {i} {j} X` is definitionally equal to `X`, so in particular it doesn't change anything to the terms of type `X`.

**Comment by identicalsnowflake** [comment]
Excellent.

This issue can be closed, although I still think that `when checking that the expression c has type Container .Agda.Primitive.lzero` is confusing since `c` does have that type.

EDIT: Never mind, this won't work. TypeInType is absolutely required since my interpreter lives in Setw, so `rewrite` won't work.

**Comment by identicalsnowflake** [comment]
Ok, after fighting with this some more, as far as I can see I am between a rock and a hard place: 

`TypeInType` goes too far: I don't want to destroy the universe hierarchy, I simply want to say "TrustMe this data type can live in `Set ℓ`."

`REWRITE` gives this functionality, but it gives it too late: I can only use it after I've declared my data type, and since it lives in `Setw`, it's rejected by Agda before I ever get a chance to `rewrite` it.

**Comment by andreasabel** [comment]
This issue needs a more detailed proposal what should be changed and how it would make sense.
